File: webstore/webstore-java/src/main/java/com/renan/webstore/business/impl/AuthServiceImpl.java
package com.renan.webstore.business.impl;

import com.renan.webstore.business.*;
import com.renan.webstore.controller.form.auth.*;
import com.renan.webstore.dto.UserDTO;
import com.renan.webstore.dto.mapper.UserDTOMapper;
import com.renan.webstore.model.AppUser;
import com.renan.webstore.persistence.repository.UserRepository;
import com.renan.webstore.util.MessageUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Random;

@Service
@RequiredArgsConstructor
public class AuthServiceImpl implements AuthService {
    
    private final UserRepository repository;
    private final PasswordEncoder passwordEncoder;
    private final UserDTOMapper userDTOMapper;
    private final ValidatorService validatorService;
    private final MailService mailService;
    
    public UserDTO login(LoginForm form) throws BusinessException {
        validateLoginForm(form);
        Optional<AppUser> userOpt = repository.findByEmail(form.getEmail().trim().toLowerCase());
        if (userOpt.isEmpty()) throw new BusinessException(
                MessageUtil.getErrorMessageUserNotFound(),
                BusinessException.BUSINESS_MESSAGE,
                AppErrorType.INVALID_INPUT
        );
        if (!passwordEncoder.matches(form.getPassword().trim(), userOpt.get().getPassword()))
            throw new BusinessException(
                    MessageUtil.getErrorMessageLogin(),
                    BusinessException.BUSINESS_MESSAGE,
                    AppErrorType.INVALID_INPUT
            );
        
        return userDTOMapper.apply(userOpt.get());
    }
    
    public void recoverPassword(RecoverPasswordForm form) throws BusinessException {
        validateRecoverPasswordForm(form);
        Optional<AppUser> userOpt = repository.findByEmail(form.getEmail().trim().toLowerCase());
        if (userOpt.isEmpty()) throw new BusinessException(
                MessageUtil.getErrorMessageUserNotFound(),
                BusinessException.BUSINESS_MESSAGE,
                AppErrorType.INVALID_INPUT
        );
        AppUser user = userOpt.get();
        user.setNewPasswordCode(generateNewPasswordCode());
        repository.save(user);
        mailService.sendNewPasswordEmail(user.getEmail(), user.getNewPasswordCode());
    }
    
    public UserDTO resetPasswordByCode(PasswordResetForm form) throws BusinessException {
        validatePasswordResetForm(form);
        Optional<AppUser> userOpt = repository.findByEmail(form.getEmail().trim().toLowerCase());
        if (userOpt.isEmpty()) throw new BusinessException(
                MessageUtil.getErrorMessageUserNotFound(),
                BusinessException.BUSINESS_MESSAGE,
                AppErrorType.INVALID_INPUT
        );
        AppUser user = userOpt.get();
        if (!user.getNewPasswordCode().equals(form.getNewPasswordCode().trim())) throw new BusinessException(
                MessageUtil.getErrorMessageInputValues(),
                BusinessException.BUSINESS_MESSAGE,
                AppErrorType.INVALID_INPUT
        );
        user.setNewPasswordCode(null);
        user.setActive(true);
        user.setPassword(passwordEncoder.encode(form.getPassword()));
        repository.save(user);
        return userDTOMapper.apply(user);
    }
    
    public void registerUser(UserRegistrationForm form) throws BusinessException {
        validateUserRegistrationForm(form);
        Optional<AppUser> userOpt = repository.findByEmail(form.getEmail().trim().toLowerCase());
        if (userOpt.isPresent()) throw new BusinessException(
                MessageUtil.getErrorMessageEmailAlreadyExists(),
                BusinessException.BUSINESS_MESSAGE,
                AppErrorType.INVALID_INPUT
        );
        AppUser newUser = new AppUser();
        newUser.setEmail(form.getEmail().trim().toLowerCase());
        newUser.setPassword(passwordEncoder.encode(form.getPassword().trim()));
        List<String> roles = new ArrayList<>();
        roles.add(AppUser.ROLE_PUBLIC);
        newUser.setRoles(newUser.buildFieldRoles(roles));
        newUser.setActive(false);
        newUser.setFirstName(form.getFirstName().trim());
        newUser.setFirstName(form.getLastName() != null ? form.getLastName().trim() : null);
        newUser.setActivationCode(generateActivationCode());
        repository.save(newUser);
    }
    
    public UserDTO activateUser(UserActivationForm form) throws BusinessException {
        validateUserActivationForm(form);
        Optional<AppUser> userOpt = repository.findByEmail(form.getEmail().trim().toLowerCase());
        if (userOpt.isEmpty()) throw new BusinessException(
                MessageUtil.getErrorMessageUserNotFound(),
                BusinessException.BUSINESS_MESSAGE,
                AppErrorType.INVALID_INPUT
        );
        AppUser user = userOpt.get();
        if (!user.getActivationCode().equals(form.getActivationCode().trim())) throw new BusinessException(
                MessageUtil.getErrorMessageInputValues(),
                BusinessException.BUSINESS_MESSAGE,
                AppErrorType.INVALID_INPUT
        );
        user.setActivationCode(null);
        user.setActive(true);
        repository.save(user);
        return userDTOMapper.apply(user);
    }
    
    public void requestUserActivation(RequestUserActivationForm form) throws BusinessException {
        validateRequestUserActivationForm(form);
        Optional<AppUser> userOpt = repository.findByEmail(form.getEmail().trim().toLowerCase());
        if (userOpt.isEmpty()) throw new BusinessException(
                MessageUtil.getErrorMessageUserNotFound(),
                BusinessException.BUSINESS_MESSAGE,
                AppErrorType.INVALID_INPUT
        );
        AppUser user = userOpt.get();
        user.setActivationCode(generateActivationCode());
        repository.save(user);
        mailService.sendActivationEmail(user.getEmail(), user.getActivationCode());
    }
    
    private void validateLoginForm(LoginForm form) throws BusinessException {
        if (form == null ||
                !validatorService.isMail(form.getEmail()) ||
                !validatorService.isValidPassword(form.getPassword()))
            throw new BusinessException(
                    MessageUtil.getErrorMessageInputValues(),
                    BusinessException.BUSINESS_MESSAGE,
                    AppErrorType.INVALID_INPUT
            );
    }
    
    private void validateRecoverPasswordForm(RecoverPasswordForm form) throws BusinessException {
        if (form == null || !validatorService.isMail(form.getEmail()))
            throw new BusinessException(
                    MessageUtil.getErrorMessageInputValues(),
                    BusinessException.BUSINESS_MESSAGE,
                    AppErrorType.INVALID_INPUT
            );
    }
    
    private void validatePasswordResetForm(PasswordResetForm form) throws BusinessException {
        if (form == null ||
                !validatorService.isMail(form.getEmail()) ||
                !validatorService.isValidPassword(form.getPassword()) ||
                !validatorService.isValidNewPasswordCode(form.getNewPasswordCode()))
            throw new BusinessException(
                    MessageUtil.getErrorMessageInputValues(),
                    BusinessException.BUSINESS_MESSAGE,
                    AppErrorType.INVALID_INPUT
            );
    }
    
    private void validateUserRegistrationForm(UserRegistrationForm form) throws BusinessException {
        if (form == null ||
                !validatorService.isMail(form.getEmail()) ||
                !validatorService.isValidPassword(form.getPassword()) ||
                form.getFirstName() == null || form.getFirstName().isBlank())
            throw new BusinessException(
                    MessageUtil.getErrorMessageInputValues(),
                    BusinessException.BUSINESS_MESSAGE,
                    AppErrorType.INVALID_INPUT
            );
    }
    
    private void validateUserActivationForm(UserActivationForm form) throws BusinessException {
        if (form == null ||
                !validatorService.isMail(form.getEmail()) ||
                !validatorService.isValidActivationCode(form.getActivationCode()))
            throw new BusinessException(
                    MessageUtil.getErrorMessageInputValues(),
                    BusinessException.BUSINESS_MESSAGE,
                    AppErrorType.INVALID_INPUT
            );
    }
    
    private void validateRequestUserActivationForm(RequestUserActivationForm form) throws BusinessException {
        if (form == null ||
                !validatorService.isMail(form.getEmail()))
            throw new BusinessException(
                    MessageUtil.getErrorMessageInputValues(),
                    BusinessException.BUSINESS_MESSAGE,
                    AppErrorType.INVALID_INPUT
            );
    }
    
    private String generateActivationCode() {
        return generateRandomString(AppUser.LENGTH_ACTIVATION_CODE);
    }
    
    private String generateNewPasswordCode() {
        return generateRandomString(AppUser.LENGTH_NEW_PASSWORD_CODE);
    }
    
    private String generateRandomString(Integer length) {
        char[] chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789".toCharArray();
        Random random = new Random();
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < length; i++) {
            int r = random.nextInt(chars.length - 1);
            result.append(chars[r]);
        }
        return result.toString();
    }
    
}

File: webstore/webstore-kotlin/src/main/kotlin/com/renan/webstore/business/impl/AuthServiceImpl.kt
package com.renan.webstore.business.impl

import com.renan.webstore.business.AppErrorType
import com.renan.webstore.business.AuthService
import com.renan.webstore.business.BusinessException
import com.renan.webstore.business.MailService
import com.renan.webstore.business.validator.Validator
import com.renan.webstore.controller.auth.form.*
import com.renan.webstore.dto.UserDTO
import com.renan.webstore.dto.mapper.UserDTOMapper
import com.renan.webstore.model.*
import com.renan.webstore.persistence.dao.UserDAO
import com.renan.webstore.util.getErrorMessageEmailAlreadyExists
import com.renan.webstore.util.getErrorMessageInputValues
import com.renan.webstore.util.getErrorMessageLogin
import com.renan.webstore.util.getErrorMessageUserNotFound
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service

@Service
class AuthServiceImpl(
    val userDAO: UserDAO,
    val mailService: MailService,
    val validator: Validator,
    val passwordEncoder: PasswordEncoder,
    val userDTOMapper: UserDTOMapper
): AuthService {

    @Throws(BusinessException::class)
    override fun login(form: LoginForm): UserDTO {
        validateLoginForm(form)
        val user = userDAO.findByEmail(form.email.trim().lowercase())
        if (user?.id == null || user.id <= 0) {
            throw BusinessException(
                message = getErrorMessageUserNotFound(),
                businessMessage = true,
                errorType = AppErrorType.INVALID_INPUT
            )
        }
        if (!passwordEncoder.matches(form.password.trim(), user.userPassword)) {
            throw BusinessException(
                message = getErrorMessageLogin(),
                businessMessage = true,
                errorType = AppErrorType.INVALID_INPUT
            )
        }
        return userDTOMapper.apply(user);
    }

    @Throws(BusinessException::class)
    override fun recoverPassword(form: RecoverPasswordForm) {
        if (!validator.isMail(form.email)) throw BusinessException(
            message = getErrorMessageInputValues(),
            businessMessage = true,
            errorType = AppErrorType.INVALID_INPUT
        )
        val user = userDAO.findByEmail(form.email.trim().lowercase())
        if (user?.id == null || user.id <= 0) {
            throw BusinessException(
                message = getErrorMessageUserNotFound(),
                businessMessage = true,
                errorType = AppErrorType.INVALID_INPUT
            )
        }
        user.newPasswordCode = generateNewPasswordCode()
        userDAO.updateUser(user)
        mailService.sendNewPasswordEmail(user.email, user.newPasswordCode!!)
    }

    @Throws(BusinessException::class)
    override fun resetPasswordByCode(form: PasswordResetForm): UserDTO {
        if (!validator.isMail(form.email) || !validator.isValidNewPasswordCode(form.newPasswordCode)) throw BusinessException(
            message = getErrorMessageInputValues(),
            businessMessage = true,
            errorType = AppErrorType.INVALID_INPUT
        )
        val user = userDAO.findByEmail(form.email.trim().lowercase())
        if (user?.id == null || user.id <= 0) {
            throw BusinessException(
                message = getErrorMessageUserNotFound(),
                businessMessage = true,
                errorType = AppErrorType.INVALID_INPUT
            )
        }
        if (user.newPasswordCode.isNullOrBlank() || !user.newPasswordCode!!.trim()
                .equals(form.newPasswordCode.trim())
        ) throw BusinessException(
            message = getErrorMessageInputValues(),
            businessMessage = true,
            errorType = AppErrorType.INVALID_INPUT
        )
        user.newPasswordCode = null
        user.userStatus = STATUS_ACTIVE
        userDAO.updateUser(user)
        return userDTOMapper.apply(user)
    }

    @Throws(BusinessException::class)
    override fun registerUser(form: UserRegistrationForm) {
        validateNewUserForm(form)
        var user = userDAO.findByEmail(form.email.trim().lowercase())
        if (user?.id != null && user.id!! > 0) {
            throw BusinessException(
                message = getErrorMessageEmailAlreadyExists(),
                businessMessage = true,
                errorType = AppErrorType.INVALID_INPUT
            )
        }
        val activationCode = generateActivationCode()
        user = AppUser(
            firstName = form.firstName.trim(),
            lastName = form.lastName?.trim(),
            email = form.email,
            userPassword = passwordEncoder.encode(form.password.trim()),
            roles = ROLE_PUBLIC,
            activationCode = activationCode,
            userStatus = STATUS_INACTIVE,
            id = null,
            newPasswordCode = null
        )
        userDAO.saveUser(user)
        mailService.sendActivationEmail(user.email, activationCode)
    }

    @Throws(BusinessException::class)
    override fun activateUser(form: UserActivationForm): UserDTO {
        if (!validator.isMail(form.email) || !validator.isValidActivationCode(form.activationCode)) throw BusinessException(
            message = getErrorMessageInputValues(),
            businessMessage = true,
            errorType = AppErrorType.INVALID_INPUT
        )
        val user = userDAO.findByEmail(form.email.trim().lowercase())
        if (user?.id == null || user.id <= 0) {
            throw BusinessException(
                message = getErrorMessageUserNotFound(),
                businessMessage = true,
                errorType = AppErrorType.INVALID_INPUT
            )
        }
        if (user.activationCode.isNullOrBlank() || !user.activationCode!!.trim().equals(form.activationCode.trim()))
            throw BusinessException(
                message = getErrorMessageInputValues(),
                businessMessage = true,
                errorType = AppErrorType.INVALID_INPUT
            )
        user.userStatus = STATUS_ACTIVE
        user.activationCode = null
        userDAO.updateUser(user)
        return userDTOMapper.apply(user)
    }

    @Throws(BusinessException::class)
    override fun requestUserActivation(form: RequestUserActivationForm) {
        if (!validator.isMail(form.email)) throw BusinessException(
            message = getErrorMessageInputValues(),
            businessMessage = true,
            errorType = AppErrorType.INVALID_INPUT
        )
        val user = userDAO.findByEmail(form.email.trim().lowercase())
        if (user?.id == null || user.id <= 0) {
            throw BusinessException(
                message = getErrorMessageUserNotFound(),
                businessMessage = true,
                errorType = AppErrorType.INVALID_INPUT
            )
        }
        user.activationCode = generateActivationCode()
        userDAO.updateUser(user)
        mailService.sendActivationEmail(user.email, user.activationCode!!)
    }

    override fun loadUserByUsername(username: String?): UserDetails {
        if (username.isNullOrBlank() || !validator.isMail(username)) throw BusinessException(
            message = getErrorMessageLogin(),
            businessMessage = true,
            errorType = AppErrorType.INVALID_INPUT
        )
        return userDAO.findByEmail(username)
            ?: throw BusinessException(
                message = getErrorMessageUserNotFound(),
                businessMessage = true,
                errorType = AppErrorType.INVALID_INPUT
            )
    }

    @Throws(BusinessException::class)
    private fun validateLoginForm(form: LoginForm) {
        if (!validator.isMail(form.email) || !validator.isValidPassword(form.password)) {
            throw BusinessException(
                message = getErrorMessageInputValues(),
                businessMessage = true,
                errorType = AppErrorType.INVALID_INPUT
            )
        }
    }

    @Throws(BusinessException::class)
    private fun validateNewUserForm(form: UserRegistrationForm) {
        if (!validator.isMail(form.email) || !validator.isValidPassword(form.password)
            || form.firstName.isEmpty()
        ) {
            throw BusinessException(
                message = getErrorMessageInputValues(),
                businessMessage = true,
                errorType = AppErrorType.INVALID_INPUT
            )
        }
    }

}